File: web/test/verify.mjs
// راستی‌آزمایی: خروجی پورت جاوااسکریپت باید بایت‌به‌بایت با پایتون یکی باشد.
import { readFileSync, readdirSync } from 'node:fs';
import { createHash } from 'node:crypto';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { setTables, decryptFileText, pretty } from '../src/lib/npv.js';

const here = path.dirname(fileURLToPath(import.meta.url));
const repo = path.resolve(here, '..', '..');
const srcDir = process.argv[2];
if (!srcDir) {
  console.error('usage: node verify.mjs <dir-with-npvt>');
  process.exit(2);
}

setTables(JSON.parse(readFileSync(path.join(repo, 'npv_tables.json'), 'utf8')));

const files = readdirSync(srcDir).filter((f) => f.toLowerCase().endsWith('.npvt'));
let ok = 0, fail = 0;
for (const f of files) {
  const text = readFileSync(path.join(srcDir, f), 'utf8');
  const blobs = decryptFileText(text);
  const joined = blobs.map((b) => pretty(b)).join('\n');
  const h = createHash('sha256').update(joined, 'utf8').digest('hex');
  if (blobs.length) { ok++; } else { fail++; }
  console.log(`${h}  ${f}  (${blobs.length} blob)`);
}
console.log(`SUMMARY ${ok} ok / ${fail} fail`);

File: verify.py
# -*- coding: utf-8 -*-
"""معادل پایتونی verify.mjs: هش خروجی هر فایل را چاپ می‌کند."""
import hashlib, os, sys
sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__))))
import decrypt_npvt as D

src = sys.argv[1]
ok = fail = 0
for f in sorted(os.listdir(src)):
    if not f.lower().endswith('.npvt'):
        continue
    parts = D.decrypt_file(os.path.join(src, f))
    joined = '\n'.join(D.pretty(p) for p in parts)
    h = hashlib.sha256(joined.encode('utf-8')).hexdigest()
    if parts:
        ok += 1
    else:
        fail += 1
    print(f'{h}  {f}  ({len(parts)} blob)')
print(f'SUMMARY {ok} ok / {fail} fail')
